snapcraft/pluginhandler.py

```diff
@@ -169,15 +169,26 @@ def _load_code(self, plugin_name, properties, part_schema):
 
         plugin = _get_plugin(module)
         options = _make_options(part_schema, properties, plugin.schema())
-        # For backwards compatibility we add the project options as an option
-        setattr(options, 'project', self._project_options)
-        self.code = plugin(self.name, options)
+        # For backwards compatibility we add the project to the plugin
+        try:
+            self.code = plugin(self.name, options, self._project_options)
+        except TypeError:
+            logger.warning('DEPRECATED: this plugin needs updating to use '
+                           'project options')
+            self.code = plugin(self.name, options)
+            # This is for plugins that don't inherit from BasePlugin
+            if not hasattr(self.code, 'project'):
+                setattr(self.code, 'project', self._project_options)
+            # This is for plugins that inherit from BasePlugin but don't have
+            # project in init.
+            if not self.code.project:
+                self.code.project = self._project_options
```

Comment:
Thanks for the comments here, very clear. We can probably make that deprecation a little more helpful-- how about something like:

```python
logger.warning('DEPRECATED: the plugin used by part {!r} needs to be updated to accept project options in its initializer. See <link to plugins.md> for more information'.format(self.name)
```

And we can add that third param to the `docs/plugins.md` document.